private/bufpkg/bufmodule/bufmoduleref/bufmoduleref.go

```diff
@@ -364,6 +383,54 @@ func ValidateModulePinsUniqueByIdentity(modulePins []ModulePin) error {
 	return nil
 }
 
+// ValidateModulePinsConsistentDigests verifies that module pins to the same commit don't change digests.
+// This is important to avoid MITM issues, where the module digest stored in a buf.lock file doesn't match
+// the module pin returned from the BSR.
+// Returns DigestChangedError if any valid digest changed from the buf.lock file for the same dependency commit.
+func ValidateModulePinsConsistentDigests(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+	modulePins []ModulePin,
+) error {
+	currentConfig, err := buflock.ReadConfig(ctx, bucket)
+	if err != nil {
+		if storage.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	if len(currentConfig.Dependencies) == 0 {
+		return nil
+	}
+	currentIdentityAndCommitToDigest := make(map[string]string, len(currentConfig.Dependencies))
+	for _, dep := range currentConfig.Dependencies {
+		// Ignore dependencies with no digest
+		if dep.Digest == "" {
+			continue
+		}
+		// Ignore dependencies with an invalid digest.
+		// We want to replace these with a valid digest.
+		if _, err := manifest.NewDigestFromString(dep.Digest); err != nil {
+			continue
+		}
```

Comment:
Worth to log a warning about replacing an invalid digest?

Reply:
We could but I don't know what value it would have. I guess we could conceivably have really old lock files that still have a b3 digest in them, and replacing it shouldn't trigger warnings.